protocol/src/main/java/org/threadly/litesockets/protocols/http/shared/HTTPHeaders.java

```diff
@@ -1,46 +1,59 @@
-package org.threadly.litesockets.protocol.http.structures;
+package org.threadly.litesockets.protocols.http.shared;
 
 import java.util.Collections;
-import java.util.HashMap;
 import java.util.Map;
 import java.util.Map.Entry;
+import java.util.TreeMap;
 
 /**
  * This is an immutable object of http headers.  These are the key/value pairs
  * separated by a colon. 
  */
 public class HTTPHeaders {
-  public final String rawHeaders;
-  public final Map<String, String> headers;
+  private final String rawHeaders;
+  private final Map<String, String> headers;
   
-  public HTTPHeaders(String rawHeaders) {
-    HashMap<String, String> map = new HashMap<String, String>();
-    this.rawHeaders = rawHeaders.intern();
-    String[] rows = rawHeaders.split(HTTPConstants.HTTP_NEWLINE_DELIMINATOR);
+  public HTTPHeaders(String headerString) {
+    TreeMap<String, String> map = new TreeMap<String, String>(String.CASE_INSENSITIVE_ORDER);
+    if(headerString.endsWith(HTTPConstants.HTTP_DOUBLE_NEWLINE_DELIMINATOR)) {
+      this.rawHeaders = headerString.substring(0, headerString.length()-2).intern();
+    } else if(!headerString.endsWith(HTTPConstants.HTTP_NEWLINE_DELIMINATOR)) { 
+      this.rawHeaders = (headerString+HTTPConstants.HTTP_NEWLINE_DELIMINATOR).intern();
+    } else {
+      while(Character.isWhitespace(headerString.charAt(0))) {
```

Comment:
yuck...if you don't want to trim the end, put this in a StringBuilder first to avoid the GC overhead